File: node_server/sim/worldEvents.js
import { EventType } from '../../shared/eventTypes.js';
export { EventType };

/**
 * Full world state snapshot, emitted once when a subscriber registers.
 */
export function worldSnapshotEvent(world) {
    const tiles = [];
    for (let x = 0; x < world.size; x++) {
        for (let y = 0; y < world.size; y++) {
            tiles.push(tileSnapshot(world.tiles[x][y]));
        }
    }
    return {
        type: EventType.WorldSnapshot,
        name: world.name,
        size: world.size,
        simTime: world.simTime,
        population: world.citizens.length,
        demand: { ...world.stats.demand },
        tiles,
    };
}

/**
 * Emitted for each tile whose state changed during a tick or command.
 */
export function tileChangedEvent(tile) {
    return {
        type: EventType.TileChanged,
        ...tileSnapshot(tile),
    };
}

/**
 * Emitted once per tick with world-level statistics.
 */
export function statsChangedEvent(world) {
    return {
        type: EventType.StatsChanged,
        simTime: world.simTime,
        population: world.citizens.length,
        demand: { ...world.stats.demand },
    };
}

function tileSnapshot(tile) {
    return {
        x: tile.x,
        y: tile.y,
        terrain: tile.terrain,
        building: tile.building ? buildingSnapshot(tile.building) : null,
    };
}

function buildingSnapshot(b) {
    const snap = {
        type: b.type,
        status: b.status,
        hideTerrain: b.hideTerrain,
        needsRoadAccess: b.needsRoadAccess,
        hasRoadAccess: b.hasRoadAccess,
    };
    if (b.name) {
        snap.name = b.name;
    }
    if (b.development) {
        snap.development = {
            state: b.development.state,
            level: b.development.level,
        };
    }
    if (b.power) {
        snap.power = {
            supplied: b.power.supplied,
            required: b.power.required,
        };
    }
    if (b.residents !== undefined) {
        snap.residents = b.residents.map(r => ({ name: r.name, age: r.age, state: r.state }));
        snap.maxResidents = b.maxResidents;
    }
    if (b.workers !== undefined) {
        snap.workers = b.workers.map(w => ({ name: w.name, age: w.age, state: w.state }));
        snap.maxWorkers = b.maxWorkers;
    }
    if (b.commerce !== undefined) {
        snap.commerce = { capacity: b.commerce.capacity };
    }
    return snap;
}

File: Simulation/Systems/WorldEvents.cs
using Simulation.Model.Buildings;
using Simulation.Model.Buildings.Zones;

namespace Simulation.Systems;

using Simulation.Model;

// ── Event type constants (wire-protocol strings) ─────────────────────────────

public static class EventType
{
    public const string WorldSnapshot = "WorldSnapshot";
    public const string TileChanged   = "TileChanged";
    public const string StatsChanged  = "StatsChanged";
}

// ── Event records ─────────────────────────────────────────────────────────────

public abstract record SimEvent(string Type);

public record WorldSnapshotEvent(
    string Name,
    int    Size,
    int    SimTime,
    int    Population,
    DemandSnapshot            Demand,
    IReadOnlyList<TileSnapshot> Tiles
) : SimEvent(EventType.WorldSnapshot);

public record TileChangedEvent(
    int             X,
    int             Y,
    string          Terrain,
    BuildingSnapshot? Building
) : SimEvent(EventType.TileChanged);

public record StatsChangedEvent(
    int            SimTime,
    int            Population,
    DemandSnapshot Demand
) : SimEvent(EventType.StatsChanged);

// ── Snapshot value objects ────────────────────────────────────────────────────

public record DemandSnapshot(double Residential, double Commercial, double Industrial);
public record TileSnapshot(int X, int Y, string Terrain, BuildingSnapshot? Building);
public record DevelopmentSnapshot(string State, int Level);
public record PowerSnapshot(int Supplied, int Required);
public record CitizenSnapshot(string Name, int Age, string State);
public record CommerceSnapshot(int Capacity);

public record BuildingSnapshot(
    string  Type,
    string  Status,
    bool    HideTerrain,
    bool    NeedsRoadAccess,
    bool    HasRoadAccess,
    string? Name          = null,
    DevelopmentSnapshot?            Development  = null,
    PowerSnapshot?                  Power        = null,
    IReadOnlyList<CitizenSnapshot>? Residents    = null,
    int?                            MaxResidents = null,
    IReadOnlyList<CitizenSnapshot>? Workers      = null,
    int?                            MaxWorkers   = null,
    CommerceSnapshot?               Commerce     = null
);

// ── Factory ───────────────────────────────────────────────────────────────────

public static class WorldEventFactory
{
    public static WorldSnapshotEvent WorldSnapshot(World world)
    {
        var tiles = new List<TileSnapshot>(world.Size * world.Size);
        for (int x = 0; x < world.Size; x++)
            for (int y = 0; y < world.Size; y++)
                tiles.Add(TileSnap(world.Tiles[x][y]));

        return new WorldSnapshotEvent(
            world.Name, world.Size, world.SimTime, world.Citizens.Count,
            DemandSnap(world.Stats.Demand), tiles);
    }

    public static TileChangedEvent TileChanged(Tile tile) =>
        new(tile.X, tile.Y, tile.Terrain,
            tile.Building is { } b ? BuildingSnap(b) : null);

    public static StatsChangedEvent StatsChanged(World world) =>
        new(world.SimTime, world.Citizens.Count, DemandSnap(world.Stats.Demand));

    // ── private helpers ───────────────────────────────────────────────────────

    private static TileSnapshot TileSnap(Tile tile) =>
        new(tile.X, tile.Y, tile.Terrain,
            tile.Building is { } b ? BuildingSnap(b) : null);

    private static DemandSnapshot DemandSnap(Demand d) =>
        new(d.Residential, d.Commercial, d.Industrial);

    private static BuildingSnapshot BuildingSnap(Building b)
    {
        var development = b is Zone z
            ? new DevelopmentSnapshot(z.Development.State, z.Development.Level)
            : null;

        var power = b.Power is { } p ? new PowerSnapshot(p.Supplied, p.Required) : null;

        IReadOnlyList<CitizenSnapshot>? residents = null;
        int? maxResidents = null;
        if (b is ResidentialZone rz)
        {
            residents    = rz.Residents.Select(r => new CitizenSnapshot(r.Name, r.Age, r.State)).ToList();
            maxResidents = rz.MaxResidents;
        }

        IReadOnlyList<CitizenSnapshot>? workers = null;
        int? maxWorkers = null;
        if (b is ZoneWithJobs zwj)
        {
            workers    = zwj.Workers.Select(w => new CitizenSnapshot(w.Name, w.Age, w.State)).ToList();
            maxWorkers = zwj.MaxWorkers;
        }

        var commerce = b is CommercialZone cz ? new CommerceSnapshot(cz.Commerce.Capacity) : null;

        return new BuildingSnapshot(
            b.Type, b.Status, b.HideTerrain, b.NeedsRoadAccess, b.HasRoadAccess,
            b.Name, development, power, residents, maxResidents, workers, maxWorkers, commerce);
    }
}
